app/code/Magento/Inventory/Model/SourceRepository.php

```diff
@@ -0,0 +1,190 @@
+    private $resourceSource;
+
+    /**
+     * @var ResourceSourceCarrierLink
+     */
+    private $resourceSourceCarrierLink;
+
+    /**
+     * @var SourceInterfaceFactory
+     */
+    private $sourceFactory;
+
+    /**
+     * @var CollectionProcessorInterface
+     */
+    private $collectionProcessor;
+
+    /**
+     * @var CollectionFactory
+     */
+    private $collectionFactory;
+
+    /**
+     * @var CarrierLinkCollectionFactory
+     */
+    private $carrierLinkCollectionFactory;
+
+    /**
+     * @var SourceSearchResultsInterfaceFactory
+     */
+    private $sourceSearchResultsFactory;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * SourceRepository constructor.
+     * @param ResourceSource $resourceSource
+     * @param ResourceSourceCarrierLink $resourceSourceCarrierLink
+     * @param SourceInterfaceFactory $sourceFactory
+     * @param CollectionProcessorInterface $collectionProcessor
+     * @param CollectionFactory $collectionFactory
+     * @param CarrierLinkCollectionFactory $carrierLinkCollectionFactory
+     * @param SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        ResourceSource $resourceSource,
+        ResourceSourceCarrierLink $resourceSourceCarrierLink,
+        SourceInterfaceFactory $sourceFactory,
+        CollectionProcessorInterface $collectionProcessor,
+        CollectionFactory $collectionFactory,
+        CarrierLinkCollectionFactory $carrierLinkCollectionFactory,
+        SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        LoggerInterface $logger
+    ) {
+        $this->resourceSource = $resourceSource;
+        $this->resourceSourceCarrierLink = $resourceSourceCarrierLink;
+        $this->sourceFactory = $sourceFactory;
+        $this->collectionProcessor = $collectionProcessor;
+        $this->collectionFactory = $collectionFactory;
+        $this->carrierLinkCollectionFactory = $carrierLinkCollectionFactory;
+        $this->sourceSearchResultsFactory = $sourceSearchResultsFactory;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function save(SourceInterface $source)
+    {
+        try {
+            /** @var SourceInterface|AbstractModel $source */
+            $this->resourceSource->save($source);
+
+            /** @var SourceCarrierLinkInterface|AbstractModel $carrierLink */
+            foreach ($source->getCarrierLinks() as $carrierLink) {
+                $carrierLink->setData(SourceInterface::SOURCE_ID, $source->getSourceId());
+                $this->resourceSourceCarrierLink->save($carrierLink);
+            }
```

Comment:
Here we have a logic problem, because we just save new sourceCarrier links, but don't remove old ones.
So, having link A and B.
We can't process change link to C
All that we can is to add C to already existing links

What we need to do , is to check that $source->getCarrierLinks is not NULL 
and if links are specified. We need to drop old links and overwrite them with new links provided in the scope of $source

```
if ($source->getCarrierLinks() !== null) {
    $connection->delete(
        $this->resourceConnection->getTableName(‘table_name_for_source_carrier_link’),
        $connection->quoteInto(‘source_id = ?‘, $source->getId())
    );
}
```